ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala

```diff
@@ -3,52 +3,340 @@
+        _ <- fetchIou(ledger, houseOwner, iou)
+        _ <- fetchIou(ledger, painter, iou).failed
+
+        offer <- ledger.create(painter, PaintOffer(painter, houseOwner, bank, onePound))
+
+        // The house owner and the painter can see the offer but the bank can't
+        _ <- fetchPaintOffer(ledger, houseOwner, offer)
+        _ <- fetchPaintOffer(ledger, painter, offer)
+        _ <- fetchPaintOffer(ledger, bank, offer).failed
+
+        // Note that the new IOU was divulged to the house owner
+        tree <- ledger.exercise(houseOwner, offer.exercisePaintOffer_Accept(_, iou))
+        (newIouEvent +: _, agreementEvent +: _) = createdEvents(tree).partition(
+          _.getTemplateId == Tag.unwrap(Iou.id))
+        newIou = Primitive.ContractId[Iou](newIouEvent.contractId)
+        agreement = Primitive.ContractId[PaintAgree](agreementEvent.contractId)
+
+        // The Bank can see the new IOU, but it cannot see the PaintAgree contract
+        _ <- fetchIou(ledger, bank, newIou)
+        _ <- fetchPaintAgree(ledger, bank, agreement).failed
+
+        // The house owner and the painter can see the contract
+        _ <- fetchPaintAgree(ledger, painter, agreement)
+        _ <- fetchPaintAgree(ledger, houseOwner, agreement)
+
+        // The painter sees its new IOU but the house owner cannot see it
+        _ <- fetchIou(ledger, painter, newIou)
+        _ <- fetchIou(ledger, houseOwner, newIou).failed
+
+      } yield {
+        // Nothing to do, all checks have been done in the for-comprehension
+      }
+    }
+
+  private def fetchIou(
+      ledger: LedgerTestContext,
+      party: Primitive.Party,
+      iou: Primitive.ContractId[Iou]): Future[Unit] =
+    for {
+      fetch <- ledger.create(party, FetchIou(party, iou))
+      _ <- ledger.exercise(party, fetch.exerciseFetchIou_Fetch)
+    } yield ()
+
+  private def fetchPaintOffer(
+      ledger: LedgerTestContext,
+      party: Primitive.Party,
+      paintOffer: Primitive.ContractId[PaintOffer]
+  ): Future[Unit] =
+    for {
+      fetch <- ledger.create(party, FetchPaintOffer(party, paintOffer))
+      _ <- ledger.exercise(party, fetch.exerciseFetchPaintOffer_Fetch)
+    } yield ()
+
+  private def fetchPaintAgree(
+      ledger: LedgerTestContext,
+      party: Primitive.Party,
+      agreement: Primitive.ContractId[PaintAgree]
+  ): Future[Unit] =
+    for {
+      fetch <- ledger.create(party, FetchPaintAgree(party, agreement))
+      _ <- ledger.exercise(party, fetch.exerciseFetchPaintAgree_Fetch)
+    } yield ()
+
+  /*
+   * Divulgence
+   */
+
+  private[this] val divulgence =
+    LedgerTest("SemanticDivulgence", "Respect divulgence rules") { ledger =>
+      for {
+        Vector(issuer, owner, delegate) <- ledger.allocateParties(3)
+        token <- ledger.create(issuer, Token(issuer, owner, 1))
+        delegation <- ledger.create(owner, Delegation(owner, delegate))
+
+        // The owner tries to delegate with a non-consuming choice, which actually doesn't
```

Comment:
typo `s/doesn't/doesn't work`

Reply:
Also `s/delegate/divulge`. Fixing, thanks.

(FYI, point 6 [here](https://help.github.com/en/articles/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request) can come in handy for this kind of comments :wink:)

Reply:
Address by https://github.com/digital-asset/daml/pull/2718/commits/d3e4ae9db78fec56118cae9c519a3aea8f728c7c

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2718/commits/d3e4ae9db78fec56118cae9c519a3aea8f728c7c